src/aspire/orientation/__init__.py

```diff
@@ -1,38 +1,503 @@
+
+        if n_equations < 2 * n_img:
+            logger.warning('Number of equations is small. Consider increase memory_factor.')
+
+        return n_equations
+
+    def _generate_shift_phase_and_filter(self, r_max, max_shift, shift_step):
+        """
+        Prepare the shift phases and generate filter for common-line detection
+
+        The shift phases are pre-defined in a range of max_shift that can be
+        applied to maximize the common line calculation. The common-line filter
+        is also applied to the radial direction for easier detection.
+        :param r_max: Maximum index for common line detection
+        :param max_shift: Maximum value of 1D shift (in pixels) to search
+        :param shift_step: Resolution of shift estimation in pixels
+        :return: shift phases matrix and common lines filter
         """
-        pass
 
-    def output(self):
+        # Number of shifts to try
+        n_shifts = int(np.ceil(2 * max_shift / shift_step + 1))
+
+        # only half of ray
+        rk = np.arange(-r_max, 0)
+
+        # Generate all shift phases
+        shifts = -max_shift + shift_step * np.arange(n_shifts)
+        shift_phases = np.exp(
+            np.outer(-2 * np.pi * 1j * rk / (2 * r_max + 1), shifts))
+        # Set filter for common-line detection
+        h = np.sqrt(np.abs(rk)) * np.exp(-np.square(rk) / (2 * (r_max / 4) ** 2))
+        return shifts, shift_phases, h
+
+    def _generate_index_pairs(self, n_equations):
         """
-        Output the 3D orientations
+        Generate two index lists for [i, j] pairs of images
         """
-        pass
+        idx_i = []
+        idx_j = []
+        for i in range(self.n_img - 1):
+            tmp_j = range(i + 1, self.n_img)
+            idx_i.extend([i] * len(tmp_j))
+            idx_j.extend(tmp_j)
+        idx_i = np.array(idx_i, dtype='int')
+        idx_j = np.array(idx_j, dtype='int')
+
+        # Select random pairs based on the size of n_equations
+        rp = np.random.choice(np.arange(len(idx_j)), size=n_equations, replace=False)
+
+        return idx_i[rp], idx_j[rp]
+
+    def _get_cl_indices(self, rotations, i, j, n_theta):
+        """
+        Get common line indices based on the rotations from i and j images
+
+        :param rotations: Array of rotation matrices
+        :param i: Index for i image
+        :param j: Index for j image
+        :param n_theta: Total number of common lines
+        :return: Common line indices for i and j images
+        """
+        # get the common line indices based on the rotations from i and j images
+        r_i = rotations[i]
+        r_j = rotations[j]
+        c_ij, c_ji = common_line_from_rots(r_i.T, r_j.T, 2 * n_theta)
+
+        # To match clmatrix, c_ij is always less than PI
+        # and c_ji may be be larger than PI.
+        if c_ij >= n_theta:
+            c_ij -= n_theta
+            c_ji -= n_theta
+        if c_ji < 0:
+            c_ji += 2 * n_theta
+
+        c_ij = int(c_ij)
```

Comment:
Depends on where to do converting. The returned values from the function `common_line_from_rots` are not `int` type. Made the change to return int type in commit 49a6474 so that we can remove these two statements.

Reply:
Cool, it makes more sense for the function that computes indices to return integers imo. Thanks.